solidity/contracts/hooks/CCIPHook.sol

```diff
@@ -0,0 +1,212 @@
+    /// @param _callData The text to be sent.
+    /// @return messageId The ID of the CCIP message that was sent.
+    function _sendMessagePayNative(
+        uint64 _destinationChainSelector,
+        address _receiver,
+        bytes memory _callData
+    )
+        internal
+        onlyAllowlistedDestinationChain(_destinationChainSelector)
+        returns (bytes32 messageId)
+    {
+        // Create an EVM2AnyMessage struct in memory with necessary information for sending a cross-chain message
+        Client.EVM2AnyMessage memory evm2AnyMessage = _buildCCIPMessage(
+            _receiver,
+            _callData,
+            address(0)
+        );
+
+        // Get the fee required to send the CCIP message
+        uint256 fees = ccip_router.getFee(_destinationChainSelector, evm2AnyMessage);
+
+        if (fees > msg.value)
+            revert NotEnoughBalance(address(this).balance, fees);
+
+        // Send the CCIP message through the router and store the returned CCIP message ID
+        messageId = ccip_router.ccipSend{value: fees}(
+            _destinationChainSelector,
+            evm2AnyMessage
+        );
+
+        // Emit an event with message details
+        emit MessageSent(
+            messageId,
+            _destinationChainSelector,
+            _receiver,
+            _callData,
+            address(0),
+            fees
+        );
+
+        // Return the CCIP message ID
+        return messageId;
+    }
+
+    /// @notice Calculates the cost in native currency for the call
+    /// @param metadata StandardHookMetadata, which contains necessary data for CCIP message formation
+    /// @param message Hyperlane Message 
+    /// @return Minimum wei required for this CCIP call
+    function quoteDispatch(
+        bytes calldata metadata,
+        bytes calldata message
+    ) public view returns (uint256) {
+        // Retrieve custom metadata, which forms the CCIP message input
+        bytes memory customMetadata = metadata.getCustomMetadata();
+        (uint64 destinationChainSelector, address receiver) = abi.decode(customMetadata, (uint64, address));
+        bytes32 _id = message.id();
+        bytes memory callData = abi.encode(_id);
+        // Create an EVM2AnyMessage struct in memory with necessary information for sending a cross-chain message
+        Client.EVM2AnyMessage memory evm2AnyMessage = _buildCCIPMessage(
+            receiver,
+            callData,
+            address(0)
+        );
+
+        // Get the fee required to send the CCIP message
+        uint256 fees = ccip_router.getFee(destinationChainSelector, evm2AnyMessage);
+
+        return fees;
+    }
+
+    function postDispatch(
+        bytes calldata metadata,
+        bytes calldata message
+    ) external payable {
+        bytes32 id = message.id();
+        require(
+            _isLatestDispatched(id),
+            "AbstractMessageIdAuthHook: message not latest dispatched"
```

Comment:
"CCIPHook: message not latest dispatched"

Reply:
done